### PR TITLE
Fix async tests

### PR DESCRIPTION
**Pull request by abidlabs**
Closes: #1999 

Note: commented out one of the tests because we think it will get fixed by #1967

### CODE REVIEW

File: gradio/examples.py [hunk 1]
```diff
@@ -198,7 +228,12 @@ def process_example(self, example_id: int) -> Tuple[List[Any], List[float]]:
             input_component.preprocess(raw_input[i])
             for i, input_component in enumerate(self.inputs)
         ]
-        predictions = self.fn(*processed_input)
+        if inspect.iscoroutinefunction(self.fn):
+            predictions = await self.fn(*processed_input)
+        else:
+            predictions = await anyio.to_thread.run_sync(
+                self.fn, *processed_input, limiter=self.limiter
+            )
```
Comment (FarukOzderim) [inline gradio/examples.py:236]: It looks like Example class is not linked to Blocks class, but could we make use of the function in Blocks?

`async def call_function(self, fn_index, processed_input):
`

File: gradio/examples.py [hunk 2]
```diff
@@ -25,6 +28,27 @@
...
+    """Top-level synchronous function that creates Examples. Provided for backwards compatibility, i.e. so that gr.Examples(...) can be used to create the Examples component."""
+    examples_obj = Examples(
+        examples=examples,
+        inputs=inputs,
+        outputs=outputs,
+        fn=fn,
+        cache_examples=cache_examples,
+        examples_per_page=examples_per_page,
+    )
+    utils.synchronize_async(examples_obj.create)
```
Comment (FarukOzderim) [inline gradio/examples.py:48]: If this does not need to be sync, we could also run it in the background with 

            utils.run_coro_in_background(fnc, parameters)

Comment (FarukOzderim) [reply gradio/examples.py:48]: Actually, let's use `run_coro_in_background`

Comment (abidlabs) [reply gradio/examples.py:50]: What if the examples take a while to cache. I think this would lead to the Interface launching without the examples caching first?

File: gradio/examples.py (hunk 1 above)
Comment (abidlabs) [reply gradio/examples.py:245]: I agree it's similar code, but it's not possible to use `call_function()` as is. We would have to rewrite it to accept a function as an argument, which I think is overkill for this situation

File: test/test_mix.py [hunk 3]
```diff
@@ -23,17 +23,17 @@ async def test_in_interface(self):
...
-        io2 = gr.Interface.load("spaces/abidlabs/image-classifier")
-        series = mix.Series(io1, io2)
-        try:
-            output = series("gradio/test_data/lion.jpg")
-            assert output["lion"] > 0.5
-        except TooManyRequestsError:
-            pass
+    # @pytest.mark.asyncio
+    # @pytest.mark.flaky
+    # async def test_with_external(self):
```
Comment (freddyaboulton) [inline test/test_mix.py:28]: How come this test was failing? And how come #1967 will fix it if it has nothing to do with sync vs async?

File: test/test_examples.py [hunk 4]
```diff
@@ -59,20 +61,22 @@ def test_some_headers(self):
 
 
 class TestProcessExamples:
```
Comment (freddyaboulton) [inline test/test_examples.py:63]: Think it would be useful to test Examples when the inference function is a coroutine?

File: gradio/examples.py [hunk 5]
```diff
@@ -55,6 +82,11 @@ def __init__(
             cache_examples: if True, caches examples for fast runtime. If True, then `fn` and `outputs` need to be provided
             examples_per_page: how many examples to show per page (this parameter currently has no effect)
         """
+        if _initiated_directly:
+            raise warnings.warn(
```
Comment (freddyaboulton) [inline gradio/examples.py:86]: I think we should raise an exception. The demo would be broken if they used `gr.examples.Examples` right?

File: gradio/examples.py (hunk 2 above)
Comment (freddyaboulton) [reply gradio/examples.py:50]: Yea agreed we should run this synchronously - like we do now

File: test/test_mix.py (hunk 3 above)
Comment (abidlabs) [reply test/test_mix.py:28]: So this test is failing because the serialization/deserialization logic is not correct anymore in `Series`. 

As part of #1967, I'm actually rewriting the serialization logic, so this will either be fixed automatically or I'll take another look at it and fix it. 

We felt it would be unnecessary effort to fix it only to change the way we do serialization.

File: gradio/examples.py [hunk 6]
```diff
@@ -198,7 +237,12 @@ def process_example(self, example_id: int) -> Tuple[List[Any], List[float]]:
             input_component.preprocess(raw_input[i])
             for i, input_component in enumerate(self.inputs)
         ]
-        predictions = self.fn(*processed_input)
+        if inspect.iscoroutinefunction(self.fn):
+            predictions = await self.fn(*processed_input)
+        else:
+            predictions = await anyio.to_thread.run_sync(
+                self.fn, *processed_input, limiter=self.limiter
```
Comment (freddyaboulton) [inline gradio/examples.py:244]: self.limiter isn't defined anywhere which will lead to exception when running a blocks api app.

```
  File "/Users/freddy/sources/gradio/gradio/examples.py", line 220, in cache_interface_examples
    prediction = await self.process_example(example_id)
  File "/Users/freddy/sources/gradio/gradio/examples.py", line 244, in process_example
    self.fn, *processed_input, limiter=self.limiter
AttributeError: 'Examples' object has no attribute 'limiter'
```

This is what I get running `blocks_inputs`

**Review by freddyaboulton — COMMENTED** [review]
@abidlabs Thanks for the fix.

I think once we fix the issue with `self.limiter` this will be good to merge!

File: test/test_mix.py (hunk 3 above)
Comment (freddyaboulton) [reply test/test_mix.py:28]: Fair point! We should try to always keep the ci on main passing though. What if someone decides to cut a release tomorrow (to test the new queue for example) and they're not aware that some of the code is broken on main?

File: gradio/examples.py (hunk 2 above)
Comment (FarukOzderim) [reply gradio/examples.py:50]: Oh shoot, I was actually going to say 

Actually, let's use synchronize_async :D

File: gradio/examples.py (hunk 6 above)
Comment (FarukOzderim) [reply gradio/examples.py:244]: This is because limiter is defined on Blocks. Btw since this is just examples, we can remove limiter support from it, not need IMO.

```suggestion
                self.fn, *processed_input
```

Comment (freddyaboulton) [reply gradio/examples.py:244]: Agreed!

File: gradio/examples.py (hunk 5 above)
Comment (abidlabs) [reply gradio/examples.py:86]: Not necessarily. They can use `ex = gr.examples.Example` as long as they remember to run `await ex.create()` afterwards. Basically just what the top-level function does.

File: test/test_examples.py (hunk 4 above)
Comment (abidlabs) [reply test/test_examples.py:63]: Yup sounds good, I'll add a test

File: gradio/examples.py (hunk 6 above)
Comment (abidlabs) [reply gradio/examples.py:244]: Thanks, will remove

File: test/test_mix.py (hunk 3 above)
Comment (abidlabs) [reply test/test_mix.py:28]: Ah I agree! But if you guys think it's all right, we can make an exception in this case, because I think it's quite unlikely a user will run into this edge case (it only happens when you use `Series` with `external`) in the time it takes for us to fix it.

**Comment by abidlabs** [comment]
Thanks for the review @freddyaboulton and @FarukOzderim! I've made the changes. I'll let the tests run and will merge shortly afterwards

File: gradio/examples.py (hunk 5 above)
Comment (freddyaboulton) [reply gradio/examples.py:86]: Actually, I just realized `raise` will actually try to raise an exception. I think you mean `warnings.warn(...)` only.

Yea that's a good point! Let's just warn instead of raising an exception.

File: test/test_mix.py (hunk 3 above)
Comment (FarukOzderim) [reply test/test_mix.py:28]: Yeah, being flexible is alright. We don't have a upcoming stable release soon as well.

File: gradio/examples.py (hunk 5 above)
Comment (abidlabs) [reply gradio/examples.py:86]: Ah good catch, will fix (sorry missed this earlier)
